### PR TITLE
feat: optimistic governor upgrade

### PR DESCRIPTION
**Pull request by Reinis-FRP**
**Motivation**

Optimistic Governor contract would benefit from integrating with Optimistic Asserter instead of current Optimistic Oracle by reducing number of contract calls and improving usability.

**Summary**

Migrates Optimistic Governor to resolve transaction proposals through Optimistic Asserter and utilizes some of its features.

**Details**

In addition to migrating to Optimistic Asserter some of notable changes include:
- Proposals are deleted automatically from Optimistic Asserter callback on disputes;
- Optimistic Governor can utilize Escalation Manager to whitelist proposers/disputers;
- Explanation is appended to the asserted claim (together with proposal hash);
- In case of Optimistic Asserter upgrade old proposals can be deleted;

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol [hunk 1]
```diff
@@ -34,59 +42,62 @@ contract OptimisticGovernor is Module, Lockable {
...
-    event ProposalDeleted(bytes32 indexed proposalHash, address indexed sender, bytes32 indexed status);
+    event ProposalExecuted(bytes32 indexed proposalHash, bytes32 indexed assertionId);
+
+    event ProposalDeleted(bytes32 indexed proposalHash, bytes32 indexed assertionId);
 
     event SetBond(IERC20 indexed collateral, uint256 indexed bondAmount);
 
     event SetCollateral(IERC20 indexed collateral);
 
-    event SetRules(string indexed rules);
```
Comment (chrismaree) [inline packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:43]: why this change.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol [hunk 2]
```diff
@@ -160,38 +177,47 @@ contract OptimisticGovernor is Module, Lockable {
...
+        require(_getIdentifierWhitelist().isIdentifierSupported(_identifier), "Identifier not supported");
         identifier = _identifier;
         emit SetIdentifier(_identifier);
     }
 
     /**
-     * @notice This pulls in the most up-to-date Optimistic Oracle.
-     * @dev If a new OptimisticOracle is added and this is run between a proposal's introduction and execution, the
+     * @notice Sets the Escalation Manager for future proposals.
+     * @param _escalationManager address of the Escalation Manager, can be zero to disable this functionality.
```
Comment (chrismaree) [inline packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:199]: nit: add a comment here about who can call this and the permissioning related to this. also might be useful to add a small detail on what the implications of the EM are in the context of the OG.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol [hunk 3]
```diff
@@ -215,39 +241,41 @@ contract OptimisticGovernor is Module, Lockable {
...
-        // Check that the proposal is not already mapped to a proposal time, i.e., is not a duplicate.
-        require(proposalHashes[proposalHash] == 0, "Duplicate proposals are not allowed");
+        // Check that the proposal is not already mapped to an assertionId, i.e., is not a duplicate.
+        require(proposalHashes[proposalHash] == bytes32(0), "Duplicate proposals not allowed");
 
-        // Map the proposal hash to the current time.
-        proposalHashes[proposalHash] = time;
+        // Check the minimum required bond and use that instead if it is greater than the bondAmount.
+        uint256 minimumBond = optimisticAsserter.getMinimumBond(address(collateral));
+        uint256 totalBond = minimumBond > bondAmount ? minimumBond : bondAmount;
```
Comment (chrismaree) [inline packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:252]: the bond logic here seems a bit strange. The point of having the `bondAmount` set within the OG it to let the owner set a bond amount that is, in theory, much larger than the OA minimum, right?

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol (hunk 1 above)
Comment (Reinis-FRP) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:43]: indexing string transforms it to its hash in the logs, so unless the observer knows the actual string this does not add much valuable information to others.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol [hunk 4]
```diff
@@ -258,101 +286,87 @@ contract OptimisticGovernor is Module, Lockable {
         // Recreate the proposal hash from the inputs and check that it matches the stored proposal hash.
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:286]: o on second thought the reason you do this is so that we dont need to store the transactions on-chain. makes sence! disregard this comment.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol (hunk 3 above)
Comment (Reinis-FRP) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:252]: yes, owner of OG might require higher amount than required by OA. But if it is below we might not want this to block proposals.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol [hunk 5]
```diff
@@ -258,101 +286,87 @@ contract OptimisticGovernor is Module, Lockable {
...
+        // Callback should only be called by the Optimistic Asserter. Address would not match in case of contract
+        // upgrade, thus try deleting the proposal through deleteProposalOnUpgrade function that should revert if
+        // address mismatch was not caused by an Optimistic Asserter upgrade.
+        if (msg.sender == address(optimisticAsserter)) {
+            // Delete the disputed proposal and associated assertionId.
+            delete proposalHashes[proposalHash];
+            delete assertionIds[assertionId];
+
+            emit ProposalDeleted(proposalHash, assertionId);
+        } else deleteProposalOnUpgrade(proposalHash);
```
Comment (chrismaree) [inline packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:360]: does this other flow not need any permissioning or entry checks on it?

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol (hunk 3 above)
Comment (Reinis-FRP) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:252]: alternative would be to check that `bondAmount` meets the minimum bond when configuring OG, but this can become obsolete if final fee gets increased.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol (hunk 5 above)
Comment (Reinis-FRP) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:360]: if `assertionId` was invalid by somebody spoofing this callback then `proposalHash` would be zero and that would revert within `deleteProposalOnUpgrade`.

While if somebody was spoofing callback with valid `assertionId` then `deleteProposalOnUpgrade` would also revert due to OA having a matching assertion with non-zero asserter.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol (hunk 3 above)
Comment (md0x) [inline packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:252]: Do you think it could make sense having a public function that does this totalBond calculation based on the minimumBond so it's easier for consumer contracts to get this number and approve the contract?
Something like `getProposalBond()` that would serve the same purpose as [gitMinimumBond](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/optimistic-asserter/implementation/OptimisticAsserter.sol#L360)

Comment (Reinis-FRP) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:252]: good idea, will add this!

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol [hunk 6]
```diff
@@ -215,39 +249,38 @@ contract OptimisticGovernor is Module, Lockable {
...
+                bytes32(0) // domainId is not set.
+            );
 
-        uint256 challengeWindowEnds = time + liveness;
+        // Maps the proposal hash to the returned assertionId and vice versa.
+        proposalHashes[proposalHash] = assertionId;
+        assertionIds[assertionId] = proposalHash;
 
-        emit TransactionsProposed(proposer, time, proposal, proposalHash, _explanation, challengeWindowEnds);
+        emit TransactionsProposed(proposer, time, assertionId, proposal, proposalHash, _explanation, time + liveness);
```
Comment (Reinis-FRP) [inline packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:283]: Do we also need emitting rules as part of TransactionsProposed? This would though increase LOC due to splitting this across multiple lines and Optimistic Asserter emits claim that will have rules appended.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol (hunk 1 above)
Comment (chrismaree) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:43]: very interesting. good thing to know!

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol (hunk 6 above)
Comment (chrismaree) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:283]: it would still be part of the same event, right? I think it's ok to emit it as well. ideally events are as expressive as possible! extra LOC for more expressiveness in events seems worth it to me.

File: packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol [hunk 7]
```diff
@@ -362,27 +381,56 @@ contract OptimisticGovernor is Module, Lockable {
...
+                AncillaryData.appendKeyValueBytes32("", PROPOSAL_HASH_KEY, _proposalHash),
+                ",",
+                EXPLANATION_KEY,
+                ':"',
+                _explanation,
+                '",',
+                RULES_KEY,
+                ':"',
+                rules,
+                '"'
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:433]: that seems good. should we not be using `ClaimData.sol` not `AncillaryData.sol`?

[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol:433]: also this makes me think that we should have a nested version that lets you pass in an array of keys and values and it builds complex longer types of concatenated data. for a future version.

**Review by chrismaree — APPROVED** [review]
This is great! the changes make sense to me and seem good! the OA does simplify a lot of things here. Thank you for other nits & cleanups that you did as well, such as better commenting and overall code hygiene.
